### ISSUE TITLE
Unknown linker for D projects on Windows

### ISSUE THREAD

**Issue by GoaLitiuM** [mesonbuild/meson#3979]
Building a simple D-program does not seem to work on Windows. With the following meson.build file:
```
project('myapp', 'd')
executable('myapp', 'myapp.d')
```
...produces the following output:
```
> meson.py build
The Meson build system
Version: 0.48.0.dev1
Source dir: C:\dev\myapp
Build dir: C:\dev\myapp\build
Build type: native build
Project name: myapp
Project version: undefined
Native D compiler: ldc2 (llvm 1.11.0-beta1 "LDC - the LLVM D compiler (1.11.0-beta1):")

meson.build:1:0: ERROR:  Unknown linker(s): [['ar']]
The follow exceptions were encountered:
Running "ar --version" gave "[WinError 2] The system cannot find the file specified"
```
On Windows platform, D compilers should be able to use following linkers, but meson does not even attempt to detect and use them:

- Digital Mars Optlink linker (link.exe), bundled with DMD and only supports 32-bit OMF targets
- Visual Studio linker (link.exe), the recommended linker for 32/64-bit MSCOFF targets
- LLVM linker (lld-link.exe), built-in linker in LDC2, and fairly recently also bundled with DMD

I tried setting the AR environment variable to point at these linkers but they all produced the same "unknown linker" error message.

Edit: I also tried running vcvarsall.bat beforehand, and it didn't help with the linker detection at all.

**Comment by nirbheek** [comment]
Yes, this is a bug, because no one has tried the D compiler support in Meson on Windows yet. Would you like to work on this? The relevant code is in `mesonbuild/environment.py` and `mesonbuild/compilers/d.py`.

It would be good to also get some tests for this on Windows. Appveyor ships with llvm on Windows, so it should be doable.

**Comment by GoaLitiuM** [comment]
Fixed as of #3981.
